File: day12.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define A(i, j) a[(i) & 255][(j) & 255]
#define R(i, j) r[(i) & 255][(j) & 255]
#define F(i, j) f[(i) & 255][(j) & 255]
char a[256][256];
int r[256][256];
bool f[256][256][4];
int n;

const int di[] = {-1, 0, 1, 0 };
const int dj[] = { 0, 1, 0,-1 };

void dfs(int i, int j, int x)
{
    R(i, j) = x;
    for (int d = 0; d < 4; d++) {
        int ni = i + di[d];
        int nj = j + dj[d];
        char c = A(ni, nj);
        if (c != A(i, j)) continue;
        if (R(ni, nj)) continue;
        dfs(ni, nj, x);
    }
}

int calc(int x, int* perim, int* fences)
{
    int area = 0;
    *perim = 0;
    *fences = 0;
    memset(f, false, sizeof(f));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (R(i, j) != x) continue;
            area++;
            for (int d = 0; d < 4; d++) {
                int ni = i + di[d];
                int nj = j + dj[d];
                if (R(ni, nj) != x) {
                    *perim += 1;
                    F(i, j)[d] = true;
                }
            }
        }
    }

    for (int d = 0; d < 4; d++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                int k = j;
                int len = 0;
                while ((d & 1 ? F(k, i) : F(i, k))[d]) {
                    k++;
                    len++;
                }
                if (len) {
                    //printf("fence %d\n", len);
                    *fences += 1;
                    j = k - 1;
                }
            }
        }
    }

    //printf("area %d perim %d fences %d\n", area, *perim, *fences);
    return area;
}

int main()
{
    FILE* file = fopen("day12.txt", "r");
    int sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%s", &A(n, 0))) {
        n++;
    }

    int x = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (R(i, j)) continue;
            dfs(i, j, ++x);
        }
    }

    for (int i = 1; i <= x; i++) {
        int perim, fences;
        int area = calc(i, &perim, &fences);
        sum += area * perim;
        sum2 += area * fences;
    }

    printf("%d %d", sum, sum2);
}

File: day12.lua
#!/usr/bin/env lua

local mt = { __index = function(t, k) t[k] = {} return t[k] end }

local a = setmetatable({}, mt)
local r = setmetatable({}, mt)
local n = 0

local di = {-1, 0, 1, 0 }
local dj = { 0, 1, 0,-1 }

local function dfs(i, j, x)
    r[i][j] = x
    local c = a[i][j]
    for d = 1, 4 do
        local ni = i + di[d]
        local nj = j + dj[d]
        if a[ni][nj] == c and not r[ni][nj] then
            dfs(ni, nj, x)
        end
    end
end

local function calc(x)
    local area = 0
    local perim = 0
    local fences = 0
    local f = setmetatable({}, mt)

    for i = 1, n do
        for j = 1, n do
            f[i][j] = {}
            if r[i][j] == x then
                area = area + 1
                for d = 1, 4 do
                    local ni = i + di[d]
                    local nj = j + dj[d]
                    if r[ni][nj] ~= x then
                        perim = perim + 1
                        f[i][j][d] = true
                    end
                end
            end
        end
    end

    for d = 1, 4 do
        for i = 1, n do
            local j = 1
            while j <= n do
                local k = j
                local len = 0
                while k <= n and (d & 1 == 0 and f[k][i] or f[i][k])[d] do
                    k = k + 1
                    len = len + 1
                end
                if len ~= 0 then
                    --print("fence", len)
                    fences = fences + 1
                    j = k - 1
                end
                j = j + 1
            end
        end
    end

    --print("area", area, "perim", perim, "fences", fences)
    return area, perim, fences
end

local file = io.open("day12.txt")
local sum = { 0, 0 }

for line in file:lines() do
    table.insert(a, { line:byte(1, #line) })
    n = n + 1
end

local x = 0
for i = 1, n do
    for j = 1, n do
        if not r[i][j] then
            x = x + 1
            dfs(i, j, x)
        end
    end
end

for i = 1, x do
    local area, perim, fences = calc(i)
    sum[1] = sum[1] + area * perim
    sum[2] = sum[2] + area * fences
end

print(sum[1], sum[2])
